src/users/users.service.ts

```diff
@@ -1,9 +1,51 @@
+// users.service.ts
+// in charge of business logic - generate slug, fetch data from other services, cache something, etc.
+import { NotFoundError } from 'elysia';
 import { UsersRepository } from '@/users/users.repository';
+import { UserInDb, UserToCreate } from '@/users/users.schema';
+import { AuthenticationError } from '@/errors';
+import { AuthService } from '@/auth/auth.service';
 
 export class UsersService {
-  constructor(private readonly repository: UsersRepository) {}
+  constructor(
+    private readonly repository: UsersRepository,
+    private readonly authService: AuthService,
```

Comment:
I don't think it's a good idea to couple auth and business logic like this. I'd much prefer to keep them separate, auth should happen before the request gets to the service

Reply:
Hmm I keep thinking about where to put the token generation part. Maybe there is no way around having the auth service in the users service without refactoring a lot of code

Reply:
In this case, upon login (in `loginUser`), the `usersService` needs to grab the user by email and then generate the token for that user. The former is the responsibility of the `usersService` (and the `usersRepository`), and the latter is in `authService` scope. `usersService` is a service, `authService` is a service. One needs to use the other for a task.

* In NestJS, services can communicate with each other. In the official guide we can see an [example that might feel close to home](https://docs.nestjs.com/fundamentals/dynamic-modules#introduction). Scrolling a bit down, we see:
  ```
  import { Injectable } from '@nestjs/common';
  import { UsersService } from '../users/users.service';
  
  @Injectable()
  export class AuthService {
    constructor(private usersService: UsersService) {}
    /*
      Implementation that makes use of this.usersService
    */
  }
  ```
* Even [in our architecture design](https://github.com/agnyz/elysia-realworld-example-app/pull/42), we outline this principle:

  > each tier communicates with its siblings or children but no child communicates with its parent.

* If this is still uncomfortable, we can look for our brothers for alternatives. That is the big benefit of this project, after all.
  * We could implement this in the framework layer, as a middleware after the information is sent back in the `usersPlugin`, or as 2 services that are managed by the `usersPlugin`. [Our Express+Firebase brothers did this](https://github.com/marcusmonteirodesouza/typescript-firestore-terraform-realworld-backend/blob/0eb57e895099fa2b39f5d5d3b5c2395ed44fffba/src/users/users-router.ts#L119).
  * We could move the token generation responsibility to the `usersService` like our [mikroORM brothers did](https://github.com/mikro-orm/nestjs-realworld-example-app/blob/3eedc3d0ba4b841323ad833c36ac7eb84952d2fa/src/user/user.service.ts#L87). Also, our [Prisma brothers](https://github.com/lujakob/nestjs-realworld-example-app/blob/034bd650cd8b9afaca77d09872f4e40b853f142a/src/user/user.service.ts#L98).

In my opinion, these are all valid options. Do you have a favorite that you'd like to implement?

Reply:
It totally makes sense, thanks! The current solution is a good one, I just like to think about stuff and consider multiple approaches. I'd say let's leave it as is and we can get back to it if for whatever reason we don't like it anymore